### PR TITLE
Issue 4 proper handling of k 1 encoding

### PR DESCRIPTION
**Pull request by maximelubin**
Encoding the single parity of k+1 code is XORing all data parts with the existing, potentially random and uninitialized content of the parity buffer.

Do a copy first, then XOR all the rest.

### CODE REVIEW

**Review by lamphamsy — APPROVED** [review]

File: binding.cc [hunk 1]
```diff
@@ -705,8 +705,16 @@ void encode(
...
     uint8_t* target = shards[flags_first(targets)];
+    int copied = 0;
     for (int i = 0; i < k + 1; i++) {
-      if (sources & (1 << i)) dot_xor(shards[i], target, shardSize);
+      if (sources & (1 << i)) {
+        if (!copied) {
+          dot_cpy(shards[i], target, shardSize);
+          copied = 1;
+        } else {
+            dot_xor(shards[i], target, shardSize);
```
Comment (jorangreef) [inline binding.cc:715]: Would you please update the indentation before `dot_xor()` to be the same as the indentation before `dot_cpy()`?

Indentation should use 2 spaces, no tabs.

File: test.js [hunk 2]
```diff
@@ -317,7 +317,7 @@ queue.onData = function(args, end) {
   );
   var buffer = cipher.update(Buffer.alloc(bufferOffset + bufferSize));
   Assert(buffer.length === bufferOffset + bufferSize);
-  var parity = Buffer.alloc(parityOffset + paritySize);
+  var parity = Node.crypto.randomBytes(parityOffset + paritySize);
```
Comment (jorangreef) [inline test.js:320]: Please would you use this instead:

`var parity = cipher.update(Buffer.alloc(parityOffset + paritySize));`

The AES cipher is faster than a call to `crypto.randomBytes()`, saving around 2 seconds (10%) of total runtime, for the same guarantees.

File: test.js [hunk 3]
```diff
@@ -360,10 +360,6 @@ queue.onData = function(args, end) {
       for (var i = 0; i < k; i++) Assert(Hash(shards[i]) === hashes[i]);
       for (var i = k; i < k + m; i++) hashes[i] = Hash(shards[i]);
       // Test against fixed vector:
-      var key = '';
```
Comment (jorangreef) [inline test.js:363]: Thanks for removing this artifact.

File: test.js [hunk 4]
```diff
@@ -764,6 +760,6 @@ queue.concat([
   [24, 5, 13,  9,      8, '4968f6d39f355c41e006b754b436d823'],
   [24, 5,  4, 12,  89816, '93e82c3652f9f963eed72e0d640d9528'],
   [24, 6, 10,  5,      8, '1cd9f45259767305dd7b565bcf9c0359'],
-  [24, 6,  3,  1, 164320, 'f2fad94fdb9e4518728cb0a5a5dbe392']
+  [24, 6,  3,  1, 164320, 'f2fad94fdb9e4518728cb0a5a5dbe392'],
```
Comment (jorangreef) [inline test.js:763]: Would you please remove the trailing comma that was added here?

**Review by jorangreef — CHANGES_REQUESTED** [review]

**Comment by jorangreef** [comment]
Thanks @maximelubin , I made the requested PR changes already in another PR:

https://github.com/ronomon/reed-solomon/commit/6f23d45c0c0a10fa5032a5d1ce6536a92c601186

**Pull request closed without merging**